File: week-03/day-2/src/Pirates/Ship.java
package Pirates;

import java.util.ArrayList;
import java.util.List;

public class Ship {
    List<Pirate> crew;
    List<Pirate> captain;
    String nameOfTheShip;
    static int shipCounter;
    int alivePiratesCounter;

    public Ship() {

        this.crew = new ArrayList<>();
        this.captain = new ArrayList<>();

        shipCounter++;
        this.nameOfTheShip = "Ship " + shipCounter;
    }

    public void fillShip() {
        captain.add(new Pirate());
        captain.get(0).promotion();
        if (Math.random() > 0.5) {
            captain.get(0).drinkSomeRum();
        }

        for (int i = 0; i < (int) (Math.random() * 1000); i++) {
            crew.add(new Pirate());
            crew.get(i).drinkSomeRum();
            crew.get(i).brawl(crew.get((int) (Math.random() * crew.size())));
        }
        alivePiratesCounter();
    }

    public int alivePiratesCounter() {
        for (Pirate memberOfCrew : crew) {
            if (memberOfCrew.dead) {
            } else {
                alivePiratesCounter++;
            }
        }
        return alivePiratesCounter;
    }


    public boolean battle(Ship otherShip) {

        if (this.alivePiratesCounter - this.captain.get(0).consumedRum > otherShip.alivePiratesCounter - otherShip.captain.get(0).consumedRum) {
            partHard(this);
            crewLost(otherShip);
            return true;
        } else {
            partHard(otherShip);
            crewLost(this);
            return false;
        }
    }

    public void partHard(Ship winningShip) {
        for (int i = 0; i < winningShip.crew.size(); i++) {
            winningShip.crew.get(i).drinkSomeRum();
        }
        winningShip.captain.get(0).drinkSomeRum();
    }

    public void crewLost(Ship looosingShip) {
        for (int i = 0; i < (int) ((int) (Math.random() * looosingShip.crew.size())); i++) {
            looosingShip.crew.get(i).die();
            alivePiratesCounter = 0;
            alivePiratesCounter();
        }
    }

    @Override
    public String toString() {
        String result = "";

        result += "Name of the Ship: " + nameOfTheShip + "\n";

        for (int i = 0; i < captain.size(); i++) {
            result += captain.get(i) + "\n";
        }

        result += "Number of alive Pirates on the ship: " + alivePiratesCounter + "\n";

        return result;
    }
}

File: week-00 (project phase)/Python/Pirates/Ship.py
from Pirate import Pirate
from random import randint

class Ship:
    def __init__(self):
        self.captain = Pirate("Captain")
        self.crew = []
        self.alive_crew = 0
        self.fill_ship()
    
    def fill_ship(self):
        self.captain.promotion()
        for pirate in range(1,randint(5,20)):
            self.crew.append(Pirate("Pirate " + str(pirate)))
        self.alive_crew = len(self.crew)

    def battle(self, challenged_ship):
        if self.alive_crew - self.captain.consumed_rum > challenged_ship.alive_crew - challenged_ship.captain.consumed_rum:
            self.party_hard()
            challenged_ship.crew_lost()
            return True
        else:
            self.crew_lost()
            challenged_ship.party_hard()
            return False

    def party_hard(self):
        for pirate in range(len(self.crew)):
            self.crew[pirate].drink_some_rum()
        self.captain.drink_some_rum()

    def crew_lost (self):
        crew_number_of_loses = randint(1, len(self.crew) - 1)
        self.alive_crew = len(self.crew) - crew_number_of_loses

        for pirate in range(randint(0, crew_number_of_loses)):
            self.crew[pirate].die
       
    def __str__(self):
        return "Captain: " + str(self.captain) + "\nAlive crew: " + str(self.alive_crew)